src/amber/pipes/cors.cr

```diff
@@ -1,81 +1,133 @@
+require "./base"
+
 module Amber
   module Pipe
-    # The CORS Handler adds support for Cross Origin Resource Sharing.
+    module Headers
+      VARY              = "Vary"
+      ORIGIN            = "Origin"
+      X_ORIGIN          = "X-Origin"
+      REQUEST_METHOD    = "Access-Control-Request-Method"
+      REQUEST_HEADERS   = "Access-Control-Request-Headers"
+      ALLOW_EXPOSE      = "Access-Control-Expose-Headers"
+      ALLOW_ORIGIN      = "Access-Control-Allow-Origin"
+      ALLOW_METHOD      = "Access-Control-Allow-Method"
+      ALLOW_HEADERS     = "Access-Control-Allow-Headers"
+      ALLOW_CREDENTIALS = "Access-Control-Allow-Credentials"
+      ALLOW_MAX_AGE     = "Access-Control-Max-Age"
+    end
+
     class CORS < Base
-      property allow_origin, allow_headers, allow_methods, allow_credentials,
-        max_age
+      alias OriginType = Array(String | Regex)
+      FORBIDDEN = "Forbidden for invalid origins, methods or headers"
+      ALLOW_METHODS = %w(PUT PATCH DELETE)
+      ALLOW_HEADERS = %w(Accept Content-type)
 
-      ALLOW_METHODS = %w(GET HEAD POST DELETE OPTIONS PUT PATCH)
-      ALLOW_HEADERS = %w(accept content-type)
+      property origins, headers, methods, credentials, max_age
+      @origin : Origin
 
       def initialize(
-        @allow_origin = "*",
-        @allow_methods = ALLOW_METHODS,
-        @allow_headers = ALLOW_HEADERS,
-        @allow_credentials = false,
-        @max_age = 0
+        @origins : OriginType = ["*", %r()],
+        @methods = ALLOW_METHODS,
+        @headers = ALLOW_HEADERS,
+        @credentials = false,
+        @max_age : Int32? = 0,
+        @expose_headers : Array(String)? = nil,
+        @vary : String? = nil
       )
+        @origin = Origin.new(origins)
       end
 
-      def initialize(
-        @allow_origin = "*",
-        allow_methods : String = ALLOW_METHODS.join(", "),
-        allow_headers : String = ALLOW_HEADERS.join(", "),
-        @allow_credentials = false,
-        @max_age = 0
-      )
-        @allow_methods = allow_methods.strip.split /[\s,]+/
-        @allow_headers = allow_headers.strip.split /[\s,]+/
+      def call(context : HTTP::Server::Context)
+        if @origin.match?(context.request)
+          put_expose_header(context.response)
+          Preflight.request?(context, self)
+          put_response_headers(context.response)
+          call_next(context)
+        else
+          return forbidden(context)
+        end
       end
 
-      def call(context : HTTP::Server::Context)
-        context.response.headers["Access-Control-Allow-Origin"] = allow_origin
+      def forbidden(context)
+        context.response.headers["Content-Type"] = "text/plain"
+        context.response.respond_with_error FORBIDDEN, 403
```

Comment:
How will this interact with the Error pipe?

Reply:
This writes to the response without raising an error. Should we raise an error here?

Reply:
I think so, that way it can be caught by the errors pipe and rendered nicely

Reply:
So we do not need to raise error here just respond with 403 according to the CORS specs since the Pref-light request is done behind the scenes by the browser